File: leetcode/lc-us/cpp/124.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    int max_sum = -1e9;

    int dfs(TreeNode *root) {
        if (!root) return 0;
        int l = max(dfs(root->left), 0), r = max(dfs(root->right), 0);
        max_sum = max(max_sum, root->val + l + r);
        return root->val + max(l, r);
    }

public:
    int maxPathSum(TreeNode *root) {
        dfs(root);
        return max_sum;
    }
};

File: leetcode/lc-us/py/124.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.res = -10 ** 9

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        def dfs(node: Optional[TreeNode]) -> int:
            if not node:
                return -10 ** 9
            left = dfs(node.left)
            right = dfs(node.right)
            self.res = max(self.res, left, right, node.val, node.val + left, node.val + right, node.val + left + right)

            return max(node.val, node.val + left, node.val + right)

        dfs(root)
        return self.res
